rfc/20250728-opentofu-providers/06a-middleware.md

```diff
@@ -0,0 +1,325 @@
+- The original operation data
+- Metadata returned by previous middleware in the chain
+- Current resource/operation context
+
+### Middleware Metadata System
+
+Middleware can attach metadata to resources that shall be persisted in both the plan and state files:
+
+#### Metadata Structure
+```json
+{
+  "__middleware_metadata__": {
+    "cost_estimator": {
+      "hourly": 0.102,
+      "monthly": 73.58,
+      "currency": "USD"
+    },
+    "approval_tracker": {
+      "approved_by": "team-lead@company.com",
+      "approval_timestamp": "2025-01-15T10:30:00Z"
+    }
+  }
+}
+```
+
+#### Metadata Requirements
+- Middleware returns an object that gets stored as `__middleware_metadata__.<MIDDLEWARENAME>`
+- Metadata is read-only and cannot modify resource configurations
+- Metadata persists across plan/apply cycles
+- Metadata is accessible to subsequent middleware in the chain
+- Metadata should be stored in both the plan file and the state file so external tooling can run against it
+
+### Middleware Response Format
+
+Middleware responds with the following data:
+
+```json
+{
+  "status": "success",  // "success", "fail", or "warning"
+  "message": "Cost estimate: $73.58/month",
+  "metadata": {
+    "estimated_cost": {
+      "hourly": 0.102,
+      "monthly": 73.58,
+      "currency": "USD"
+    }
+  }
+}
+```
+
+**Response Options**:
+- **`success`**: Allow operation to continue
+- **`fail`**: Block operation and return error
+- **`warning`**: Allow operation but display warning
+
+### Error Handling and Failures
+
+When middleware returns a "fail" status:
+- The operation is immediately halted
+- The error message is displayed to the user
+- No subsequent middleware in the chain executes
+- The resource operation is not performed
+
+### Configuration Model
+
+Middleware is provided by providers and configured using a provider-based model similar to regular providers.
+
+#### Configuration Syntax
+
+**Middleware Declaration**: `middleware "providername" "middlewarename" { }`
+- `providername` must be declared in `required_providers`
+- `middlewarename` is a local name for this middleware instance
+- Configuration block contains middleware-specific settings
+
+**Global Middleware**: `terraform { middleware = [...] }`
+- Runs for all providers and resources
+- Configured in the `terraform` block alongside `required_providers`
+
+**Provider-Specific Middleware**: `provider "name" { middleware = [...] }`
+- Runs only for resources from that specific provider
+- Configured in the provider block
+
+## Implementation Considerations
+
+### Sensitive Values and Security
+
+#### Sensitive Data Handling
+
+Middleware interaction with sensitive values could be controlled through configuration:
+
+```hcl
+middleware "audit" "logger" {
+  log_destination = "splunk"
+  
+  # Sensitive value handling
+  send_sensitive = false    # Don't send sensitive values to middleware
+  sanitize_logs = true     # Remove sensitive data from logs
```

Comment:
Can you expand on this?  I don't think we have anything like it and implementing it correctly worries me.